src/modules/olMap/services/RtVectorLayerService.js

```diff
@@ -0,0 +1,151 @@
+ * Service that creates an ol layer from a RtVectorGeoResource (Websocket)
+ * and applies specific stylings if required.
+ * @class
+ * @author thiloSchlemmer
+ */
+export class RtVectorLayerService {
+	_addPortToUrl(url, port) {
+		const pathArray = url.split('.');
+		const applyOnLastElement = (pathElements, port) => {
+			const lastElement = pathElements[pathElements.length - 1];
+			if (lastElement.indexOf('/') !== -1) {
+				const lastElementSegments = lastElement.split('/');
+				lastElementSegments[0] = lastElementSegments[0] + ':' + port;
+				return lastElementSegments.join('/');
+			} else {
+				return pathArray[pathArray.length - 1] + ':' + port;
+			}
+		};
+		pathArray[pathArray.length - 1] = applyOnLastElement(pathArray, port);
+		return pathArray.join('.');
+	}
+
+	_getFeatureReader(rtVectorGeoResource) {
+		const { MapService: mapService } = $injector.inject('MapService');
+		const destinationSrid = mapService.getSrid();
+
+		const format = mapVectorSourceTypeToFormat(rtVectorGeoResource.sourceType);
+		return (data) =>
+			format
+				.readFeatures(data)
+				.filter((f) => !!f.getGeometry())
+				.map((f) => {
+					f.getGeometry().transform('EPSG:' + rtVectorGeoResource.srid, 'EPSG:' + destinationSrid);
+					return f;
+				});
+	}
+
+	/**
+	 * processes the messages from a websocket and updates the specified olVectorSource
+	 *
+	 * TODO: check, if behavior for changed featureExtent is needed
+	 * @param {MessageEvent} socketEvent
+	 * @param {ol.source.Vector} olVectorSource
+	 * @param {function (string): Array<ol.Feature>} featureReader
+	 */
+	_processMessage(socketEvent, olVectorSource, featureReader) {
+		const { data } = socketEvent;
+		if (data === WebSocket_Message_Keep_Alive) {
+			return;
+		}
+
+		olVectorSource.clear();
+		const features = featureReader(data);
+		olVectorSource.addFeatures(features);
+	}
+
+	_startWebSocket(rtVectorGeoResource, olVectorLayer, olMap, port) {
+		const { VectorLayerService: vectorLayerService } = $injector.inject('VectorLayerService');
+		const featureReader = this._getFeatureReader(rtVectorGeoResource);
+		const olVectorSource = olVectorLayer.getSource();
+		const webSocket = new WebSocket(port ? this._addPortToUrl(rtVectorGeoResource.url) : rtVectorGeoResource.url);
+
+		webSocket.onmessage = (event) => {
+			this._processMessage(event, olVectorSource, featureReader);
+			vectorLayerService.sanitizeStyles(olVectorLayer);
+			if (rtVectorGeoResource.isClustered()) {
+				vectorLayerService.applyClusterStyle(olVectorLayer);
+			} else {
+				vectorLayerService.applyStyles(olVectorLayer, olMap);
+			}
+		};
+
+		const tryNextPort = (port) => {
+			webSocket.onmessage = undefined;
+			const nextPort = getNextPort(WebSocket_Ports, port);
+			this._startWebSocket(rtVectorGeoResource, olVectorLayer, olMap, nextPort);
+		};
+
+		const failure = () => {
+			throw new UnavailableGeoResourceError('Realtime-data cannot be displayed for technical reasons.', rtVectorGeoResource.id);
+		};
```

Comment:
Is this the untestable part of the code?

Reply:
yes

Reply:
now refactored, to be testable